File: productor_consumer/producer_consumer.js
var i = 0

async function producer() {
    while (true) {
        if (i >= 5) {
            break
        }
        i ++
        console.log("produce:" + i)
        await consumer()
    }
}

async function consumer() {
    while (true) {
        if (i >= 5) {
            break
        }
        console.log("consume:" + i)
        await producer()
    }
}

producer().catch(error => console.log(error.stack));

File: productor_consumer/producer_consumer.lua
local newProducer

function producer()
     local i = 0
     while true do
          if i >= 5 then
               break
          end
          i = i + 1
          print(string.format("produce:%d", i))
          send(i)
     end
end

function consumer()
     while true do
          local i = receive()
          print(string.format("consume:%d", i))
          if i >= 5 then
               break
          end
     end
end

function receive()
     local status, value = coroutine.resume(newProducer)
     return value
end

function send(x)
     coroutine.yield(x)
end

newProducer = coroutine.create(producer)
consumer()